python/cuml/cluster/kmeans.pyx

```diff
@@ -432,19 +432,22 @@ class KMeans(Base):
         Sum of squared distances of samples to their closest cluster center.
         """
 
+        self._check_output_type(X)
```

Comment:
This behavior is a little funny to me. I never expect "predict" to set anything on `self`. I think it would make sense for check_output_type returns a local and does not set self, so every call just mirrors its inputs nicely. 
 For `fit`, the output_type affects fields like `coef_`, so I think only `fit` methods actually should set self.output_type. If we reset the self.output_type every time someone calls predict, then it will get out of sync with self.fields that were computed during fit.

Reply:
This method sets the output type for each method. So, assuming default output_type behavior for example, if a user does fit with a cupy array then the labels will be cupy, but if then he calls predict with a numpy array then this just makes it that the return predictions from predict are numpy arrays. Internally all is cuml arrays (cuArrays?) so they are not affected except at the last return statement. If the user wants a consistent return format, there is the option of setting output_type at the estimator or module levels

Reply:
Deleted my comment here, since on further analysis and running things, I think you're right. Only fit (and similar) methods should affect the output_type of the estimator when it is in mirroring input mode. Predict/transform methods shouldn't affect the output_type of things like the labels and owned structures of the estimator. Will do the change accordingly.